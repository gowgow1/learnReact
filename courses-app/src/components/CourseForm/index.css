.create-course {
	padding: 15px 20px;
	border: 2px solid lightskyblue;
}

.create-course .button{
	padding: 5px 5px;
	border-radius: 0;
	border: 1px solid rgb(255, 0, 255);
}

.create-course .input {
	margin: 0 0 15px;
}

.control {
	display: flex;
	justify-content: space-between;
	margin-bottom: 15px;
}



.text-aria {
	width: 100%;
	min-height: 200px;
	margin-bottom: 15px;
}

.author-wrap {
	display: flex;
	column-gap: 50px;
}


.author-create, .author-choose {
	flex: 1 1 auto;
	width: 50%;
	justify-content: center;
}

.author-name {
	display: flex;
	justify-content: space-between;
	margin-bottom: 10px;
}

.duration-bar span {
	font-size: 20px;
	font-weight: 600;
}

h2 {
	font-size: 20px;
	text-align: center;
}

ul {
	padding-left: 0;
}


