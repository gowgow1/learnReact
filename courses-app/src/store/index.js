import { createStore, combineReducers } from 'redux';
import courses from './courses/reducer';
import user from './user/reducer';

const store = createStore(combineReducers({ courses, user }));

export default store;
