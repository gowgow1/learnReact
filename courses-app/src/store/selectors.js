export const getUser = ({ user }) => user;
export const getCourses = ({ courses }) => courses;
export const getAuthors = ({ authors }) => authors;
