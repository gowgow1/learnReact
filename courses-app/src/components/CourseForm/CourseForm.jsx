import { useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import Input from '../common/Input';
import Button from '../common/Button';

import timeFormat from '../../helpers/timeFormat';
import { getAuthors, getCourses, getUser } from '../../store/selectors';
import { asyncUpdateCourse, asyncAddCourse } from '../../store/courses/thunk';
import { asyncAddAuthor } from '../../store/authors/thunk';

import './index.css';

const CourseForm = () => {
	const { courseId } = useParams();
	const history = useHistory();

	const dispatch = useDispatch();
	const initialAuthors = useSelector(getAuthors);
	const initialCourses = useSelector(getCourses);

	const { token } = useSelector(getUser);
	const course = initialCourses.find((item) => item.id === courseId);

	const [currentAuthors, setAuthors] = useState([]);
	const [choosenAuthors, setChoosenAuthors] = useState([]);
	const [title, setTitle] = useState('');
	const [description, setDescription] = useState('Some text...');
	const [authorName, setAuthorName] = useState('');
	const [duration, setDuration] = useState(0);

	const initStates = (course) => {
		if (course) {
			const choosenAuthors = initialAuthors.filter((author) =>
				course.authors.includes(author.id)
			);
			const authors = initialAuthors.filter(
				(author) => !course.authors.includes(author.id)
			);
			setAuthors(authors);
			setChoosenAuthors(choosenAuthors);
			setTitle(course.title);
			setDescription(course.description);
			setDuration(course.duration);
		} else {
			setAuthors(initialAuthors);
		}
	};

	useEffect(() => {
		initStates(course);
	}, [initialAuthors, initialCourses]);

	const addTitle = (event) => setTitle(event.target.value);
	const addDescription = (event) => setDescription(event.target.value);
	const addAuthorName = (event) => setAuthorName(event.target.value);
	const addDuration = (event) => setDuration(+event.target.value);

	const onAddAuthor = (event, id) => {
		setChoosenAuthors([
			...choosenAuthors,
			currentAuthors.find((item) => item.id === id),
		]);

		setAuthors(currentAuthors.filter((item) => item.id !== id));
		event.preventDefault();
	};

	const onRemoveAuthor = (event, id) => {
		setAuthors([
			...currentAuthors,
			choosenAuthors.find((item) => item.id === id),
		]);
		setChoosenAuthors(choosenAuthors.filter((item) => item.id !== id));

		event.preventDefault();
	};

	const updateAuthors = (event) => {
		event.preventDefault();
		dispatch(asyncAddAuthor({ name: authorName }, token));
	};

	const updateCourse = (event) => {
		event.preventDefault();
		const date = new Date();
		dispatch(
			asyncUpdateCourse(token, {
				title,
				description,
				duration,
				creationDate: `${date.getDate()}/${date.getMonth()}/${date.getFullYear()}`,
				authors: choosenAuthors.map((item) => item.id),
				id: course.id,
			})
		);
		history.push('/courses');
	};

	const setCourse = (event) => {
		event.preventDefault();
		const date = new Date();
		dispatch(
			asyncAddCourse(token, {
				title,
				description,
				duration,
				creationDate: `${date.getDate()}/${date.getMonth()}/${date.getFullYear()}`,
				authors: choosenAuthors.map((item) => item.id),
			})
		);
		history.push('/courses');
	};

	return (
		<form className='create-course'>
			<div className='control'>
				<Input
					placeholder='Enter Title'
					labelText='Title'
					id='title'
					onChange={addTitle}
					name='title'
					value={title}
				/>
				{courseId ? (
					<Button text='Update course' onClick={updateCourse} />
				) : (
					<Button text='Create course' onClick={setCourse} />
				)}
			</div>
			<label htmlFor='ariaText'>Description</label>
			<textarea
				minLength='2'
				id='ariaText'
				className='text-aria'
				value={description}
				onChange={addDescription}
				name='description'
			/>
			<div className='author-wrap'>
				<div className='author-create'>
					<h2>Add author</h2>
					<Input
						placeholder='Enter author name...'
						labelText='Author name:'
						id='authorInput'
						name='authorInput'
						onChange={addAuthorName}
						value={authorName}
					/>
					<Button text='Create Author' onClick={updateAuthors} />
					<h2>Duration</h2>
					<Input
						placeholder='Enter duration in minutes...'
						labelText='Duration'
						id='duration'
						onChange={addDuration}
						name='duration'
						type='number'
						value={duration}
					/>
					<div className='duration-bar'>
						Duration: <span>{timeFormat(duration)}</span> hours
					</div>
				</div>
				<div className='author-choose'>
					<h2> Authors </h2>
					<ul>
						{currentAuthors.map(({ id, name }) => (
							<li key={id} className='author-name'>
								<div>{name}</div>
								<Button
									text='Add author'
									onClick={(event) => onAddAuthor(event, id)}
								/>
							</li>
						))}
					</ul>
					<h2>Course authors</h2>
					{choosenAuthors.length ? (
						<ul>
							{choosenAuthors.map(({ id, name }) => (
								<li key={id} className='author-name'>
									<div>{name}</div>
									<Button
										text='Delete Author'
										onClick={(event) => onRemoveAuthor(event, id)}
									/>
								</li>
							))}
						</ul>
					) : (
						'Author list is empty'
					)}
				</div>
			</div>
		</form>
	);
};

CourseForm.propTypes = {
	initialAuthors: PropTypes.array,
	updateAuthors: PropTypes.func,
	setCourse: PropTypes.func,
};

export default CourseForm;
