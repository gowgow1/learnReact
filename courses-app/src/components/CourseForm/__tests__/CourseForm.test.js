import { Provider } from 'react-redux';
import * as reactRedux from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';

import CourseForm from '../CourseForm';
import * as selectors from '../../../store/selectors';
import { mockedAuthorsList, mockedCoursesList } from '../../../constants';

const mockedState = {
	user: {
		isAuth: true,
		name: 'alex',
		token: 'fasdasdaf123',
		role: 'admin',
		email: 'admin@email.ru',
	},
	courses: mockedCoursesList,
	authors: mockedAuthorsList,
};

const mockedStore = {
	getState: () => mockedState,
	subscribe: jest.fn(),
	dispatch: jest.fn(),
};

jest.mock('react-redux', () => ({
	...jest.requireActual('react-redux'),
	useDispatch: jest.fn(),
}));

jest.mock('../../../store/selectors', () => ({
	getCourses: jest.fn(),
	getAuthors: jest.fn(),
	getUser: jest.fn(),
}));

window.alert = jest.fn();

const useDispatchSpyOn = jest.spyOn(reactRedux, 'useDispatch');

beforeEach(() => {
	selectors.getAuthors.mockReturnValue(mockedState.authors);
	selectors.getCourses.mockReturnValue(mockedState.courses);
	selectors.getUser.mockReturnValue(mockedState.user);
	reactRedux.useDispatch.mockReturnValue(jest.fn());
});

afterEach(() => {
	selectors.getAuthors.mockClear();
	selectors.getCourses.mockClear();
	selectors.getUser.mockClear();
	reactRedux.useDispatch.mockClear();
});

describe('CourseForm', () => {
	it('should be rendered', () => {
		const { container } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<CourseForm />
				</BrowserRouter>
			</Provider>
		);
		expect(container).toMatchSnapshot();
	});

	it('should show authors', () => {
		const { queryAllByTestId } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<CourseForm />
				</BrowserRouter>
			</Provider>
		);
		expect(queryAllByTestId('authors')).toHaveLength(mockedAuthorsList.length);
	});

	it('"createAuthor" should call dispatch', () => {
		const { getByTestId } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<CourseForm />
				</BrowserRouter>
			</Provider>
		);

		fireEvent.click(getByTestId('createAuthor'));
		expect(useDispatchSpyOn).toHaveBeenCalled();
	});

	it('"Add author" add an author in author list', () => {
		const { getAllByTestId, getByTestId } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<CourseForm />
				</BrowserRouter>
			</Provider>
		);
		const addAuthor = getAllByTestId('addAuthor');
		fireEvent.click(addAuthor[0]);
		expect(getByTestId('choosenAuthor')).toBeTruthy();
	});

	it('"Delete Author" should delte author from current authors list', () => {
		const { getAllByTestId, getByTestId } = render(
			<Provider store={mockedStore}>
				<BrowserRouter>
					<CourseForm />
				</BrowserRouter>
			</Provider>
		);
		const addAuthors = getAllByTestId('addAuthor');
		expect(addAuthors).toHaveLength(mockedAuthorsList.length);
		fireEvent.click(addAuthors[0]);
		expect(getAllByTestId('addAuthor')).toHaveLength(
			mockedAuthorsList.length - 1
		);
		fireEvent.click(getByTestId('deleteAuthor'));
		expect(getAllByTestId('addAuthor')).toHaveLength(mockedAuthorsList.length);
	});
});
