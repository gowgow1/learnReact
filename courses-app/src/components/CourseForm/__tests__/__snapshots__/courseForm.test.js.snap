// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseForm should be rendered 1`] = `
<div>
  <form
    class="create-course"
  >
    <div
      class="control"
    >
      <div
        class="input-wrap"
      >
        <label
          for="title"
        >
          Title
        </label>
        <input
          class="input"
          id="title"
          name="title"
          placeholder="Enter Title"
          type="text"
          value=""
        />
      </div>
      <button
        class="button"
      >
        Create course
      </button>
    </div>
    <label
      for="ariaText"
    >
      Description
    </label>
    <textarea
      class="text-aria"
      id="ariaText"
      minlength="2"
      name="description"
    >
      Some text...
    </textarea>
    <div
      class="author-wrap"
    >
      <div
        class="author-create"
      >
        <h2>
          Add author
        </h2>
        <div
          class="input-wrap"
        >
          <label
            for="authorInput"
          >
            Author name:
          </label>
          <input
            class="input"
            id="authorInput"
            name="authorInput"
            placeholder="Enter author name..."
            type="text"
            value=""
          />
        </div>
        <button
          class="button"
          data-testid="createAuthor"
        >
          Create Author
        </button>
        <h2>
          Duration
        </h2>
        <div
          class="input-wrap"
        >
          <label
            for="duration"
          >
            Duration
          </label>
          <input
            class="input"
            id="duration"
            name="duration"
            placeholder="Enter duration in minutes..."
            type="number"
            value="0"
          />
        </div>
        <div
          class="duration-bar"
        >
          Duration: 
          <span>
            00:00
          </span>
           hours
        </div>
      </div>
      <div
        class="author-choose"
      >
        <h2>
           Authors 
        </h2>
        <ul>
          <li
            class="author-name"
            data-testid="authors"
          >
            <div>
              Vasiliy Dobkin
            </div>
            <button
              class="button"
              data-testid="addAuthor"
            >
              Add author
            </button>
          </li>
          <li
            class="author-name"
            data-testid="authors"
          >
            <div>
              Nicolas Kim
            </div>
            <button
              class="button"
              data-testid="addAuthor"
            >
              Add author
            </button>
          </li>
          <li
            class="author-name"
            data-testid="authors"
          >
            <div>
              Anna Sidorenko
            </div>
            <button
              class="button"
              data-testid="addAuthor"
            >
              Add author
            </button>
          </li>
          <li
            class="author-name"
            data-testid="authors"
          >
            <div>
              Valentina Larina
            </div>
            <button
              class="button"
              data-testid="addAuthor"
            >
              Add author
            </button>
          </li>
        </ul>
        <h2>
          Course authors
        </h2>
        Author list is empty
      </div>
    </div>
  </form>
</div>
`;
